tests/kafka/test_kafka.py

```diff
@@ -0,0 +1,232 @@
+    ts = None
+
+    topic = _random_name("topic")
+    _await(kafka_admin.create_topics([NewTopic(topic, num_partitions=1)]))
+
+    for i in range(3):
+        key = str(i)
+        kafka_producer.produce(topic, b"value" + key.encode(), key.encode())
+        kafka_producer.flush()
+
+        if i == 0:
+            ts = pendulum.now()
+
+        time.sleep(5)
+
+    yield topic, ts
+
+    _await(kafka_admin.delete_topics([topic]))
+
+
+def test_kafka_read(kafka_topics, kafka_messages):
+    """Test simple messages reading."""
+    pipeline = dlt.pipeline(
+        pipeline_name="kafka_test",
+        destination="postgres",
+        dataset_name="kafka_test_data",
+        full_refresh=True,
+    )
+
+    resource = kafka_consumer(kafka_topics)
+    load_info = pipeline.run(resource)
+
+    assert_load_info(load_info)
+
+    table_names = [t["name"] for t in pipeline.default_schema.data_tables()]
+    table_counts = load_table_counts(pipeline, *table_names)
+
+    assert set(table_counts.keys()) == set(kafka_topics)
+
+    for tab in table_counts:
+        assert table_counts[tab] == 3
+
+    for tab in table_names:
+        assert_query_data(
+            pipeline,
+            f"SELECT _kafka__key FROM {tab} ORDER BY _kafka__key",
+            sorted(kafka_messages[tab]),
+        )
+
+
+def test_kafka_read_custom_msg_processor(kafka_topics, kafka_messages):
+    """
+    Test messages reading and processing with a
+    custom processor-function.
+    """
+
+    def _custom_msg_processor(msg):
+        return {
+            "_kafka": {
+                "topic": msg.topic(),
+            },
+            "value": "test!",
+        }
+
+    pipeline = dlt.pipeline(
+        pipeline_name="kafka_test",
+        destination="postgres",
+        dataset_name="kafka_test_data",
+        full_refresh=True,
+    )
+    resource = kafka_consumer(kafka_topics, msg_processor=_custom_msg_processor)
+    load_info = pipeline.run(resource)
+
+    assert_load_info(load_info)
+
+    table_names = [t["name"] for t in pipeline.default_schema.data_tables()]
+
+    for tab in table_names:
+        assert_query_data(
+            pipeline,
+            f"SELECT value FROM {tab}",
+            ["test!", "test!", "test!"],
+        )
+
+
+def test_kafka_read_with_timestamp(kafka_timed_messages):
+    """Test reading, starting from a particular timestamp."""
+    topic, ts = kafka_timed_messages
+
+    pipeline = dlt.pipeline(
+        pipeline_name="kafka_test",
+        destination="postgres",
+        dataset_name="kafka_test_data",
+        full_refresh=True,
+    )
+    resource = kafka_consumer(topic, start_from=ts)
+    load_info = pipeline.run(resource)
+
+    assert_load_info(load_info)
+
+    assert_query_data(
+        pipeline,
+        f"SELECT _kafka__key FROM {topic} ORDER BY _kafka__key",
+        ["1", "2"],
+    )
+
+
+def test_kafka_incremental_read(kafka_producer, kafka_topics):
```

Comment:
Okay, I made the test more complex. The thing is - it can have flaky errors. For now, I didn't get them in these tests, but I when I was testing manually, it happened several times. For example, a topic is not yet replicated, by a consumer is already trying to read messages from it, using an unreplicated broker. Replication and balancing should be reduced on the cluster. We're not reading data eternally (as it was designed to be in Kafka), but extract a fixated amount of data - replication will only create problems in this case.